### PR TITLE
Fix CognitiveComplexity to not throw an exception on Lucene.Net

### LINKED ISSUE #727: Fix CognitiveComplexity to not throw an exception on Lucene.Net
The analyzer `SonarAnalyzer.Rules.CSharp.CognitiveComplexity` throws an exception of type `System.InsufficientExecutionStackException` when analyzing `Lucene.Net` on peach.
```
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
SC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.MetricsAnalyzer' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
CSC : warning AD0001: Analyzer 'SonarAnalyzer.Rules.CSharp.CognitiveComplexity' threw an exception of type 'System.InsufficientExecutionStackException' with message 'Insufficient stack to continue executing the program safely. This can happen from having too many functions on the call stack or function on the stack using too much stack space.'. [d:\slave02\workspace\Dotnet_lucene.net-GP6PSKNHI45IDLU4FS53BQR3UUICZ6Q6NMHCQBJYSLEL4EFUX6SQ\src\Lucene.Net.Analysis.Common\Lucene.Net.Analysis.Common.csproj]
```

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #727

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs [hunk 1]
```diff
@@ -51,6 +53,25 @@ public void EnsureVisitEndedCorrectly()
...
 
+        public void Walk(SyntaxNode node)
+        {
+            try
+            {
+                Visit(node);
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                // TODO: trace this exception
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs [hunk 2]
```diff
@@ -31,6 +31,8 @@ public sealed class CognitiveComplexityWalker : CSharpSyntaxWalker
     {
         private readonly List<SecondaryLocation> incrementLocations = new List<SecondaryLocation>();
         private readonly List<ExpressionSyntax> logicalOperationsToIgnore = new List<ExpressionSyntax>();
+        private readonly Queue<BinaryExpressionSyntax> binaryNodesToVisit =
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: MAJOR Remove the unused private field 'binaryNodesToVisit'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs [hunk 3]
```diff
@@ -51,6 +53,25 @@ public void EnsureVisitEndedCorrectly()
...
+            catch (InsufficientExecutionStackException)
+            {
+                // TODO: trace this exception
+
+                // Roslyn walker overflows the stack when the depth of the call is around 2050.
+                // See ticket #727.
+
+                // Reset nesting level, so the problem with the walker is not reported.
+                nestingLevel = 0;
+                return;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: MINOR Remove this redundant jump. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3626)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: Please remove this unused field.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs [hunk 4]
```diff
@@ -51,6 +51,24 @@ public void EnsureVisitEndedCorrectly()
...
 
+        public void Walk(SyntaxNode node)
+        {
+            try
+            {
+                Visit(node);
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                // TODO: trace this exception
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs (hunk 4 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CognitiveComplexityWalker.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**